File: Data_structure/BOJ_10866.cpp
#include <iostream>
#include <algorithm>
#include <string>
#include <deque>


using namespace std;


deque<int> dq;

int main()
{
	int req, num;
	string str;
	cin >> req;
	for(int i = 0; i < req; i++){
		cin >> str;
		if(!str.compare("push_front")){
			cin >> num;
			dq.push_front(num);
		}
		if(!str.compare("push_back")){
			cin >> num;
			dq.push_back(num);
		}
		if(!str.compare("pop_front")){
			if(dq.empty()){
				cout << "-1\n"; 
				continue;
			}
			cout << dq.front() << endl;
			dq.pop_front();
		}
		if(!str.compare("pop_back")){
			if(dq.empty()){
				cout << "-1\n"; 
				continue;
			}
			cout << dq.back() << endl;
			dq.pop_back();
		}
		if(!str.compare("size")){
			cout <<dq.size() << endl;
		}
		if(!str.compare("empty")){
			cout << dq.empty() << endl;
		}
		if(!str.compare("front")){
			if(dq.empty()){
				cout << "-1\n"; 
				continue;
			}
			cout << dq.front() << endl;
		}
		if(!str.compare("back")){
			if(dq.empty()){
				cout << "-1\n";
				continue;
			}
			cout << dq.back() << endl;
		}
	}
}

File: BOJ_prob/Data_structure/python/BOJ_10866.py
import sys

if __name__ == '__main__':
    numofinst = int(sys.stdin.readline())
    listlikedeque = []
    for i in range(numofinst):
        inst = sys.stdin.readline().split()
        if inst[0] == 'push_front':
            op = int(inst[1])
            listlikedeque.insert(0, op)
        elif inst[0] == 'push_back':
            op = int(inst[1])
            listlikedeque.append(op)
        elif inst[0] == 'pop_front':
            if len(listlikedeque) == 0:
                print(-1)
            else:
                poppedval = listlikedeque.pop(0)
                print(poppedval)
        elif inst[0] == 'pop_back':
            if len(listlikedeque) == 0:
                print(-1)
            else:
                poppedval = listlikedeque.pop()
                print(poppedval)
        elif inst[0] == 'size':
            print(len(listlikedeque))
        elif inst[0] == 'empty':
            if len(listlikedeque) == 0:
                print(1)
            else:
                print(0)
        elif inst[0] == 'front':
            if len(listlikedeque) == 0:
                print(-1)
            else:
                print(listlikedeque[0])
        else:
            if len(listlikedeque) == 0:
                print(-1)
            else:
                print(listlikedeque[len(listlikedeque) - 1])


           